### ISSUE TITLE
detect shared printer model on client computers

### ISSUE THREAD

**Issue by GreatBahram** [apple/cups#5580]
Hi,
I know operating systems usually detect printer on certain network protocol such as `IPP`, `LPD`, `SNMP` and so on. My problem is I shared a printer via `cups` and computers cannot detect the printer model, they're just using it like a shared queue without any model specification.

Could you possibly explain is there any way to add a functionality and be able to detect printer model? Like by asking cups what are the printer attributes? does it support `duplex`? ...

Clients have to find the drivers manually and this sometimes can be so frustrating specially on windows operating system since without knowing printer model it cannot download the proper driver from its database.

**Comment by michaelrsweet** [comment]
What operating systems are your clients using? Any Linux/UNIX-based client won't need this information since it can get it directly from CUPS using IPP. Windows clients use the make and model information supplied by CUPS (either with IPP or via Samba/SMB).

**Comment by GreatBahram** [comment]
Sorry for the delay, I'm not sure about that. Consider an ubuntu OS, for example, when I want to add an `ipp` shared printer, after searching, it'll add the printer, it works file. But it does not know about the make and the model (It seems more like a raw queue, am I right?). When I query the ubuntu system via `python3-pycups` package, I get below result:
```python
{'printer': {'printer-is-shared': False,
  'printer-state': 3,
  'printer-state-message': '',
  'printer-state-reasons': ['none'],
  'printer-type': 2228230,
  'printer-uri-supported': 'ipp://localhost/printers/printer',
  'printer-location': '',
  'printer-info': 'HP LaserJet P2055',
  'device-uri': 'ipp://192.168.16.206:631/printers/HP-LaserJet-P2055',
  'printer-make-and-model': 'Remote Printer'}}
```
By the way, I checked the cups server, I has any requests from that client.

**Comment by GreatBahram** [comment]
`cupsd.conf` configuration
```
LogLevel debug
PageLogFormat 
MaxLogSize 0
# Allow remote access
Port 631
Listen /var/run/cups/cups.sock
# Share local printers on the local network.
Browsing Yes
#Browsing On
BrowseLocalProtocols dnssd
DefaultAuthType Basic
WebInterface No
<Location />
  # Allow shared printing...
  Order allow,deny
  Allow all
</Location>
<Location /admin>
</Location>
<Location /admin/conf>
  AuthType Default
  Require user @SYSTEM
</Location>
<Location /admin/log>
  AuthType Default
  Require user @SYSTEM
  Order allow,deny
</Location>
<Policy default>
  JobPrivateAccess default
  JobPrivateValues none
  SubscriptionPrivateAccess default
  SubscriptionPrivateValues default
  <Limit Create-Job Print-Job Print-URI Validate-Job>
    Order deny,allow
  </Limit>
  <Limit Send-Document Send-URI Hold-Job Release-Job Restart-Job Purge-Jobs Set-Job-Attributes Create-Job-Subscription Renew-Subscription Cancel-Subscription Get-Notifications Reprocess-Job Cancel-Current-Job Suspend-Current-Job Resume-Job Cancel-My-Jobs Close-Job CUPS-Move-Job CUPS-Get-Document>                                                                                                                                                                                                                                      
    Require user @OWNER @SYSTEM
    Order deny,allow
  </Limit>
  <Limit CUPS-Add-Modify-Printer CUPS-Delete-Printer CUPS-Add-Modify-Class CUPS-Delete-Class CUPS-Set-Default CUPS-Get-Devices>
    AuthType Default
    Require user @SYSTEM
    Order deny,allow
  </Limit>
  <Limit Pause-Printer Resume-Printer Enable-Printer Disable-Printer Pause-Printer-After-Current-Job Hold-New-Jobs Release-Held-New-Jobs Deactivate-Printer Activate-Printer Restart-Printer Shutdown-Printer Startup-Printer Promote-Job Schedule-Job-After Cancel-Jobs CUPS-Accept-Jobs CUPS-Reject-Jobs>                                                                                                                                                                                                                                    
    AuthType Default
    Require user @SYSTEM
    Order deny,allow
  </Limit>
  <Limit Cancel-Job CUPS-Authenticate-Job>
    Require user @OWNER @SYSTEM
    Order deny,allow
  </Limit>
  <Limit All>
    Order deny,allow
  </Limit>
</Policy>
<Policy authenticated>
  JobPrivateAccess default
  JobPrivateValues default
  SubscriptionPrivateAccess default
  SubscriptionPrivateValues default
  <Limit Create-Job Print-Job Print-URI Validate-Job>
    AuthType Default
    Order deny,allow
  </Limit>
  <Limit Send-Document Send-URI Hold-Job Release-Job Restart-Job Purge-Jobs Set-Job-Attributes Create-Job-Subscription Renew-Subscription Cancel-Subscription Get-Notifications Reprocess-Job Cancel-Current-Job Suspend-Current-Job Resume-Job Cancel-My-Jobs Close-Job CUPS-Move-Job CUPS-Get-Document>                                                                                                                                                                                                                                      
    AuthType Default
    Require user @OWNER @SYSTEM
    Order deny,allow
  </Limit>
  <Limit CUPS-Add-Modify-Printer CUPS-Delete-Printer CUPS-Add-Modify-Class CUPS-Delete-Class CUPS-Set-Default>
    AuthType Default
    Require user @SYSTEM
    Order deny,allow
  </Limit>
  <Limit Pause-Printer Resume-Printer Enable-Printer Disable-Printer Pause-Printer-After-Current-Job Hold-New-Jobs Release-Held-New-Jobs Deactivate-Printer Activate-Printer Restart-Printer Shutdown-Printer Startup-Printer Promote-Job Schedule-Job-After Cancel-Jobs CUPS-Accept-Jobs CUPS-Reject-Jobs>                                                                                                                                                                                                                                    
    AuthType Default
    Require user @SYSTEM
    Order deny,allow
  </Limit>
  <Limit Cancel-Job CUPS-Authenticate-Job>
    AuthType Default
    Require user @OWNER @SYSTEM
    Order deny,allow
  </Limit>
  <Limit All>
    Order deny,allow
  </Limit>
</Policy>
```

**Comment by GreatBahram** [comment]
`printers.conf` content:
```xml
<Printer HP-LaserJet-300-400-color-MFP-M375-M475>
UUID urn:uuid:349c8471-0a3a-32d4-47f3-ef9d87e8d623
Info HP LaserJet 300-400 color MFP M375-M475
Location B-Team
MakeModel HP LJ 300-400 color MFP M375-M475 Postscript (recommended)
DeviceURI custom_backend://HP-LaserJet-300-400-color-MFP-M375-M475
State Idle
StateTime 1558183099
ConfigTime 1558183099
Type 8425564
Accepting Yes
Shared Yes
JobSheets none none
QuotaPeriod 0
PageLimit 0
KLimit 0
OpPolicy default
ErrorPolicy retry-job
</Printer>
<Printer HP-LaserJet-P2055>
UUID urn:uuid:147a9a03-e11f-3cbb-5f40-54d7025f8472
Info HP LaserJet P2055
Location NSG Team
MakeModel HP LaserJet P2055 Postscript (recommended)
DeviceURI custom_backend://HP-LaserJet-P2055
State Idle
StateTime 1558183080
ConfigTime 1558183080
Type 8425556
Accepting Yes
Shared Yes
JobSheets none none
QuotaPeriod 0
PageLimit 0
KLimit 0
OpPolicy default
ErrorPolicy retry-job
</Printer>
```

**Comment by jsmeix** [comment]
@GreatBahram 
Perhaps I do not fully understand what you mean
but I think you confuse print queues (some kind of "software")
with printer devices (the hardware that actually makes dots on paper).

What is "inside" CUPS are only print queues but never actual printers
so CUPS can only provide information about its print queues,
cf. "CUPS: The server between user and printer" at
https://en.opensuse.org/Concepts_printing

Some print queue information can contains the make and model
of a real printer device but that are just some text fields that
may or may not contain information about the real printer
make and model that is "behind" the print queue,
cf. "The Printer Configurations Dialog" at
https://en.opensuse.org/YaST_Printer
(regardless that it is about the YaST printer module
the general information therein applies everywhere).

Accordingly there is no such thing as
```
detect shared printer model on client computers
```
because CUPS never shares a printer device.

Regarding "Printing from Windows to Linux" you may have a look at
https://en.opensuse.org/SDB:Printing_from_Windows_to_Linux

**Comment by GreatBahram** [comment]
> What is "inside" CUPS are only print queues but never actual printers so CUPS can only provide information about its print queues.

You are certainly right. CUPS only knows about the queue. What I want is when I shared a printer via `CUPS`, `CUPS` just propagates that information. @michaelrsweet believes cups does that at the moment. But in my case at least neither Unix OS nor windows OS cannot detect the printer.

> Accordingly there is no such thing, because CUPS never shares a printer device.

I do know about cups, it's just a queue and so on. But `CUPS` knows about make and model because it reads the information from the `PPD` file. I was wonder is there any mechanism or facility to share that information with client computers or not?

By the way thanks for your links

**Comment by jsmeix** [comment]
@GreatBahram 
thank you for your explanation.
Now I think I understand what you are looking for.

As far as I understand what you are missing in
https://github.com/apple/cups/issues/5580#issuecomment-493677912
(excerpts)
```
'printer-location': '',
'printer-info': 'HP LaserJet P2055',
'device-uri': 'ipp://192.168.16.206:631/printers/HP-LaserJet-P2055',
'printer-make-and-model': 'Remote Printer'
```
is that `printer-make-and-model` should not be `Remote Printer`
but something like what `printer-info` is.

I don't know the details behind but I think there is a reason
(I even have an offhanded guess what that reason might be)
why CUPS reports its print queue as a generic `Remote Printer`
and not as if it was a real printer device like `HP LaserJet P2055`.

The source code is at
https://github.com/apple/cups/blob/master/scheduler/printers.c#L4997
but without explanatory comments one cannot know the reason
behind why the code is as it is - also neither
https://github.com/apple/cups/blob/master/cups/cups.h#L204
nor
https://www.cups.org/doc/cupspm.html
explain what CUPS_PRINTER_REMOTE actually means.

My offhanded guess what that reason might be is that usually
it would be just wrong to do a local printer setup on a client system
for a remote CUPS queue because a printer setup on a client system
usually means that print job data processing software with printer driver
software is run on the client (i.e. "filtering" happens on the client)
which produces the final printer model specific data so that
printer model specific data would be sent from the client to
the CUPS queue where that data would be again processed
by the filtering that is set up for the CUPS queue but usually
the filtering that is set up for the CUPS queue would fail or
produce a mess so that the final output on the actual printer
device would fail or result a mess, cf. the section about
"Differences in Printing between Windows and Linux" in
https://en.opensuse.org/SDB:Printing_from_Windows_to_Linux

Simply put:
Duplicated print job data processing ("filtering") must be avoided.

**Comment by GreatBahram** [comment]
Thanks for your explanation.

It's true on some client systems when they chose **inappropriate driver**, even on Unix machines, there **were some problems and sometimes even jobs failed** ( I think that was just because of different filtering system that each driver and operating system use), However, that's totally OK in my situation.

And you are right @michaelrsweet, CUPS shows information about  printer make and model, the problem is `system-config-printer` program on Linux does not read that information and I think that's the case with Windows too.

https://github.com/zdohnal/system-config-printer/blob/83856b9f4f37fe0df0ec23ced15b6f62109f64e2/probe_printer.py#L490
